src/main/kotlin/com/hedvig/accountservice/schedulers/domain/sagas/SubscriptionScheduledSaga.kt

```diff
@@ -24,190 +27,295 @@ import java.time.Instant
+    @EndSaga
+    fun on(event: ScheduleSubscriptionBackfillEvent, @Timestamp now: Instant) {
+        val currentPeriod = YearMonth.from(now.atZone(ZoneId.of("UTC")).toLocalDate())
+        val checkAllPeriods = !subscriptionAlreadyScheduled(event.memberId, firstPeriodBackfill)
+
+        if (subscriptionAlreadyScheduled(event.memberId, currentPeriod) && !checkAllPeriods) {
+            logger.info { "ALREADY SCHEDULED" }
+            return
+        }
+
+        try {
+            checkSubscriptionsOfPeriod(event, currentPeriod, event.memberId, checkAllPeriods, now)
+        } catch (exception: Exception) {
+
+            logger.error { "Schedule subscription of PERIOD (${currentlyCheckingPeriod.month} ${currentlyCheckingPeriod.year}) failed for memberId=${event.memberId} (Message: ${exception.message})" }
+            return
+        }
+
     }
 
-    commandGateway.sendAndWait<SubscriptionScheduledCommand>(
-      SubscriptionScheduledCommand(
-        event.memberId,
-        event.scheduledBy
-      )
-    )
-  }
+    private fun checkSubscriptionsOfPeriod(
+        eventType: Any,
+        period: YearMonth,
+        memberId: String,
+        checkAllPeriods: Boolean,
+        now: Instant
+    ) {
+        currentlyCheckingPeriod = period
+        //TODO: For 2019-05 this is always false in staging. Check with Elvin if we need to insert an account entry if
+        // exists in subscription_scheduled already
+        if (!subscriptionAlreadyScheduled(memberId, period)) {
+            val productSubscriptionsOfPeriod = getProductSubscriptions(memberId, period)
+            val discountOfPeriod = getCampaignDiscounts(memberId, period)
 
-  private fun checkSubscriptionsOfPeriod(
-    event: ScheduleSubscriptionEvent,
-    period: YearMonth,
-    memberId: String,
-    checkAllPeriods: Boolean,
-    now: Instant
-  ) {
-    currentlyCheckingPeriod = period
+            productSubscriptionsOfPeriod.stream().forEach({o -> logger.info { "${o.productId} -- ${o.subscription}" }})
+            scheduleProductSubscriptions(memberId, period, eventType, productSubscriptionsOfPeriod, now)
+            scheduleCampaignDiscounts(memberId, period, eventType, discountOfPeriod, now)
+
+            //Don't schedule subscriptions for backfills
+            if (eventType is ScheduleSubscriptionEvent) {
+                addSubscriptionScheduled(eventType.memberId, period, eventType.scheduledBy, now)
+            }
+        }
+
+        val previousPeriod = period.minusMonths(1)
+        if (shouldCheckPeriod(previousPeriod, memberId, checkAllPeriods, eventType)) {
+            checkSubscriptionsOfPeriod(eventType, previousPeriod, memberId, checkAllPeriods, now)
+        }
+    }
 
-    if (!subscriptionAlreadyScheduled(memberId, period)) {
-      val productSubscriptionsOfPeriod = getProductSubscriptions(memberId, period)
+    private fun scheduleProductSubscriptions(
+        memberId: String,
+        period: YearMonth,
+        event: Any,
```

Comment:
Something like this would probably be better since we don't want to have different logic for the different event types.

```suggestion
        stopAtPeriod: YearMonth,
```